Improve Error messages, & add lints for common patterns

rcoh opened the pull request:
![Selection_001](https://user-images.githubusercontent.com/492903/82250054-cdd87000-9918-11ea-973f-5bed9cbc14d8.png)

This diff adds linting machinery, but also a bunch of other stuff:
1. BooleanExpressions now track Span information to trace them back to input Yaml. This will help give better errors messages as well.
2. Adds a lint for `$A = X` and `return X` in the same pattern either
3. Adds a pretty printer for errors (utilizing the Span machinery)
4. Move a bunch of existing errors to pretty errors
5. Don't print escape sequences when the output is not a tty

I can break it up if this is too much.

Closes #678

rcoh committed:
Use attrs instead of NamedTuple to fix missing call to validate

rcoh committed:
Use attrs instead of NamedTuple to fix missing call to validate

rcoh committed:
Move attrs to setup.py

rcoh committed:
Move attrs to setup.py

rcoh committed:
Backout version.py changes

rcoh committed:
Backout version.py changes

rcoh committed:
Add machinery to compare patterns

rcoh committed:
Add machinery to compare patterns

rcoh committed:
Linters WIP

rcoh committed:
Linters WIP

rcoh committed:
Cleanup & add nicer output

rcoh committed:
Cleanup & add nicer output

rcoh committed:
Get tests passing

rcoh committed:
Get tests passing

rcoh committed:
Archive yaml.safe_load warning (we are using SafeLoader)

rcoh committed:
Archive yaml.safe_load warning (we are using SafeLoader)

rcoh committed:
Don't print escape sequences to files

rcoh committed:
Don't print escape sequences to files

rcoh committed:
Add another nice error

rcoh committed:
Add another nice error

rcoh committed:
Undo edits to version.py

rcoh committed:
Undo edits to version.py

nbrahms commented on .bento/archive.json:
```diff
@@ -1,6 +1,14 @@
+        "path": "semgrep/semgrep/config_resolver.py",
+        "message": "Use of unsafe yaml load. Allows instantiation of arbitrary objects. Consider yaml.safe_load().",
+        "severity": 1,
+        "syntactic_context": "return {config_id: yaml.load(contents, Loader=SafeLineLoader)}"
```
we really need to port bandit to get rid of this nonsense

brendongo replied:
why aren't we using safe_load?

rcoh replied:
This is actually a good check! But to get line numbers I had to go around the loader. Very tricky check to not error on what I did...

rcoh replied:
See the code. safe_load is just `load(..., loader=SafeLoader)`. We have a custom loader which tracks line numbers. I think we should switch to `ruamel` but that's a project for another day

DrewDennison replied:
+100 for ruamel. I use it exclusively for semgrep.live

brendongo commented on semgrep/semgrep/cli.py:
```diff
@@ -78,6 +81,13 @@ def cli() -> None:
     )
+
+    config.add_argument(
+        "--lint",
```
Should we have all this functionality under --validate?

rcoh replied:
I wanted to keep this separate as (for the time being) something more experimental because:
- It's kind of slow
- Some of my lints might be wrong
- It will fire on lots of existing rules in semgrep-rules currently

rcoh replied:
I could definitely imagine making it part of validate (or promoting individual lints to validations)

brendongo replied:
what's an example of when it fires on semgrep-rules?

rcoh replied:
```
lint: redundant patterns
  --> ../../semgrep-rules/python/flask/security/unsanitized_input.yaml:29
29 |           - pattern: |
30 |               $X = request.args.get(...)
31 |               ...
32 |               return flask.make_response(f"{$X}...")
  --> ../../semgrep-rules/python/flask/security/unsanitized_input.yaml:33
33 |           - pattern: |
34 |               $X = request.args.get(...)
35 |               ...
36 |               $R = flask.make_response(f"{$X}...")
= help: remove the first and delete `return` from the second
These patterns are redundant (return paired with assignment)
```

rcoh replied:
Fixed lint errors in semgrep-rules: https://github.com/returntocorp/semgrep-rules/pull/522

brendongo replied:
Idk if I would say these are redundant?

rcoh replied:
Having those two rules in combination matches what you would match with:
```
34 |               $X = request.args.get(...)
35 |               ...
36 |               flask.make_response(f"{$X}...")
```
(you no longer need to specify `return` or `$A =` explicitly.)

brendongo commented on semgrep/tests/e2e/snapshots/test_rule_parser/test_rule_parser__failure__error_messages/bad1/error.txt:
```diff
@@ -1,2 +1,14 @@
+7 |         - patterns:
+8 |           - pattern-inside: |
+9 |               def __init__(...):
+10 |                   ...
```
Can we align the `|`

rcoh replied:
yeah, I had put it off for v0 but it isn't hard to fix.

rcoh replied:
fixed.

brendongo commented on semgrep/semgrep/semgrep_types.py:
```diff
@@ -76,6 +129,10 @@ class BooleanRuleExpression:
     operand: Optional[str] = None
 
+    # For tests, eg. don't force people to make spans
+    span: Optional[Span] = None
```
Why does the BooleanRuleExpression need Span objects?

brendongo replied:
I think it would be less cognitive load to pull out the validation code out of the object into its own function/class and have that thing handle provenance.

rcoh replied:
because at the time of validation, we've lost the original source. In general, keeping spans attached to things is good because it means you can always show the relevant code for errors. (eg. if an individual booleanexpression causes a semgrep-core error)

I could be convinced we should be keeping an external mapping from `BooleanExpression` => `Span` though if you prefer

brendongo replied:
Might be enough to just rename to provenance or something to avoid confusion wrt line ranges of findings and line ranges of the rule provenance. I think if it's not much more work to pull out the validation stuff might be less messy.

brendongo commented on semgrep/semgrep/semgrep_types.py:
```diff
@@ -87,19 +144,31 @@ def _validate(self) -> None:
             if self.children is not None:
-                raise InvalidRuleSchema(
-                    f"only {pattern_names_for_operators(OPERATORS_WITH_CHILDREN)} operators can have children, but found `{pattern_names_for_operator(self.operator)}` with children"
+                from semgrep.error import SemgrepLangError
```
Why are these being imported inline?

rcoh replied:
circular import

rcoh replied:
removed.

brendongo commented on semgrep/semgrep/error.py:
```diff
@@ -18,3 +24,34 @@ class OutdatedPythonException(SemgrepException):
         self.msg = "Semgrep requires Python 3.5+. Please ensure you are using Python3.5+ to run semgrep."
+
+
+class SemgrepLangError(NamedTuple):
```
Something like `RuleYamlError` would be more clear

brendongo commented on semgrep/semgrep/semgrep_types.py:
```diff
@@ -60,13 +69,57 @@ class OPERATORS:
     pass
 
 
+class Span(NamedTuple):
```
This will be confusing with Range class that is used for findings.

rcoh replied:
moved & referenced via the namespace

brendongo commented on semgrep/semgrep/semgrep_types.py:
```diff
@@ -41,9 +42,17 @@ class OPERATORS:
     OPERATORS.REGEX: ["pattern-regex"],
 }
 
+START_LINE = "__line__"
```
Don't these only matter in parsing the yaml? Can this be defined where the custom parser is defined?

brendongo commented on semgrep/semgrep/pattern_lints.py:
```diff
@@ -0,0 +1,112 @@
+
+
+def check_equivalent_patterns(
+    pattern_either: BooleanRuleExpression, lang: str = "python"
```
let's not default to python

rcoh replied:
whoops :+1: good catch, thought I deleted all of those.

brendongo reviewed (changes requested):
See above comments

rcoh committed:
Refactorings & Cleanups

- Refactor yaml parsing into parse_yaml_with_spans
- Refactor Span and RuleLangError into rule_lang.py
- Fix alignment when the number of line numbers changes
- Remove xfail from tests (fixed cause of underlying failure)

rcoh committed:
Refactorings & Cleanups

- Refactor yaml parsing into parse_yaml_with_spans
- Refactor Span and RuleLangError into rule_lang.py
- Fix alignment when the number of line numbers changes
- Remove xfail from tests (fixed cause of underlying failure)

rcoh reviewed (commented):
Addressed everything except pulling Spans out of BooleanExpressions

rcoh committed:
Remove unused imports

rcoh committed:
Remove unused imports

rcoh committed:
Fix test output

rcoh committed:
Fix test output

rcoh committed:
Fix Pipefile.lock after rebase

rcoh committed:
Fix Pipefile.lock after rebase

rcoh commented:
```
lint: redundant patterns
  --> ../../semgrep-rules/python/flask/security/unsanitized_input.yaml:29
29 |           - pattern: |
30 |               $X = request.args.get(...)
31 |               ...
32 |               return flask.make_response(f"{$X}...")
  --> ../../semgrep-rules/python/flask/security/unsanitized_input.yaml:33
33 |           - pattern: |
34 |               $X = request.args.get(...)
35 |               ...
36 |               $R = flask.make_response(f"{$X}...")
= help: remove the first and delete `return` from the second
These patterns are redundant (return paired with assignment)
```

On Tue, May 19, 2020 at 1:07 PM Brendon Go <notifications@github.com> wrote:

> *@brendongo* commented on this pull request.
> ------------------------------
>
> In semgrep/semgrep/cli.py
> <https://github.com/returntocorp/semgrep/pull/773#discussion_r427461359>:
>
> > @@ -78,6 +81,13 @@ def cli() -> None:
>          action="store_true",
>          help="Validate configuration file(s). No search is performed.",
>      )
> +
> +    config.add_argument(
> +        "--lint",
>
> what's an example of when it fires on semgrep-rules?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/returntocorp/semgrep/pull/773#discussion_r427461359>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADYKZ4ZRPQXSVCH424HNODRSK4DJANCNFSM4NDDSU6A>
> .
>

rcoh committed:
Cleanup output & imports

rcoh committed:
Cleanup output & imports

rcoh committed:
Update for changed API

rcoh committed:
Update for changed API

rcoh committed:
Refactor, rename to provenance, fix exception causing early bail out

rcoh committed:
Refactor, rename to provenance, fix exception causing early bail out

rcoh committed:
Hide the flag entirely

rcoh committed:
Hide the flag entirely

rcoh committed:
Disable docker push from fork

rcoh committed:
Merge branch 'russell/pattern-equivalence' of github.com:rcoh/semgrep into russell/pattern-equivalence

rcoh committed:
Fix tests

rcoh committed:
Merge branch 'develop' of github.com:returntocorp/sgrep into russell/pattern-equivalence

rcoh committed:
Skip pushing to docker for forks

rcoh committed:
Fix pff version

rcoh committed:
Rollback changes to version.py

brendongo commented on CHANGELOG.md:
```diff
@@ -5,6 +5,8 @@ This project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html
 ## Next Release
 
 ### Added
+- Improved error messages for common yaml rule syntax errors
```
I'm surprised this didn't get a conflict.

brendongo commented on semgrep/semgrep/rule_match.py:
```diff
@@ -6,6 +6,16 @@
+    if isinstance(obj, list):
+        return [strip_internal_keys(o) for o in obj]
+    elif isinstance(obj, dict):
+        return {k: v for k, v in obj.items() if k not in YAML_INTERNAL_KEYS}
```
Were YAML internal keys being added to RuleMatch objects?

brendongo replied:
Afaik RuleMatches are made from PatternMatch objects/semgrep-core output and this PR doesnt change what semgrep-core outputs?

rcoh replied:
the metadata field from rules got the line numbers attached. (metadata is included in rule output)

brendongo replied:
Oh I see

rcoh commented:
I think I'm going to rebuild this on top of ruamel. Should be much cleaner and more complete
